docs/rfcs/blocks-in-widget-areas.md

```diff
@@ -0,0 +1,535 @@
+
+[sidebars_widgets]: https://developer.wordpress.org/reference/hooks/sidebars_widgets/
+
+### `dynamic_sidebar()`
+
+Usage of the [`dynamic_sidebar()`][dynamic_sidebar] function remains as-is, though it will now render any blocks in the specified widget area.
+
+There are some nuances that surround how `dynamic_sidebar()` handles the `'before_widget'`, `'after_widget'`, `'before_title'` and `'after_title'` attributes that are associated with the widget area.
+
+The `'before_title'` and `'after_title'` attributes are rendered:
+
+- Before/after the title of a `core/legacy-widget` block is rendered.
+- Never, when a non-`core/legacy-widget` block is rendered.
+
+The `'before_widget'` and `'after_widget'` attributes are rendered:
+
+- Before/after a `core/legacy-widget` block is rendered.
+- Before/after a _sequence_ of non-`core/legacy-widget` blocks are rendered.
+
+This is done so that all blocks aren't wrapped with `'before_widget'` and `'after_widget'` markup. Instead, **sequences** of blocks are combined together _and then_ wrapped with `'before_widget'` and `'after_widget'` markup. This makes blocks render on old themes as if they were all together in a single widget.
+
+To illustrate, let's say that the `'footer'` widget area is registered so that titles are wrapped in `<h2>` elements and widgets in `<section>` elements.
+
+```php
+register_sidebar( array( 
+	'id'            => 'footer',
+	...
+	'before_widget' => '<section>',
+	'after_widget'  => '</section>',
+	'before_title'  => '<h2>',
+	'after_title'   => '</h2>',
+) );
+```
+
+The `'footer'` widget area has been migrated.
+
+```php
+array(
+	'footer' => 123,
+	'array_version' => 4,
+)
+```
+
+The referenced `wp_area` contains some blocks and a legacy widget in its `post_content`.
+
+```html
+<!-- wp:paragraph --><p>Hello there!</p><!-- /wp:paragraph -->
+<!-- wp:legacy-widget {"identifier":"search-2","instance":{"title":"Search"}} /-->
+<!-- wp:separator --><hr class="wp-block-separator"/><!-- /wp:separator -->
+<!-- wp:paragraph --><p>Welcome to my cool website!</p><!-- /wp:paragraph -->
+```
+
+Now, let's call `dynamic_sidebar( 'footer' )`.
+
+```html
+<section>
+	<p>Hello there!</p>
+</section>
+<section>
+	<h2>Search</h2>
+	<form role="search" method="get" class="search-form" action="http://wordpress.test/">
+		...
+	</form>
+</section>
+<section>
+	<hr class="wp-block-seperator" />
+	<p>Welcome to my cool website!</p>
+</section>
+```
+
+[dynamic_sidebar]: https://developer.wordpress.org/reference/functions/dynamic_sidebar/
+[register_sidebar]: https://developer.wordpress.org/reference/functions/register_sidebar/
+
+## Performance
+
+`wp_area` posts will be pre-emptively fetched when rendering a page. That is, all of the posts referenced in `'sidebar_widgets'` will be fetched using a single `SELECT ... FROM wp_posts WHERE ID in ?` query. This ensures that we do not introduce one extra query per widget area.
```

Comment:
Currently on widgets areas are autoloaded on every page. Which is a currently issue. All wp_area posts would need to load on every page, as you just never know what widget areas are displayed on each page.

Reply:
That's a good point. It's tricky to know ahead of time which widget areas will be used, though. We'd have to (ab)use output buffering or some such.

I think that this is an optimisation we should look at doing separately in the future.

Reply:
We have two ways of going about this, either always every wp_area post on every request, or just load them as they are requests. So one big request or lots of little ones. Lots of little ones makes more sense, say some pages don't have widgets on them at all, like say a 404 page...

Reply:
What's more common: That a template uses all of the sidebars registered by the theme, or that it doesn't?

Reply:
There is no common way of using widgets areas, it is just unknownable. I feel like we are going to have load all widget areas on every request, which I hate but I dont know if they there is anyway of knowing which widget areas in use at anytime.